### PR TITLE
Feature/90 retain email log table

### PR DESCRIPTION
**Pull request by mariadanieldeepak**
This fixes https://github.com/sudar/email-log/issues/89

### CODE REVIEW

File: include/Core/UI/Setting/EmailLogSetting.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
+<?php
+namespace EmailLog\Core\UI\Setting;
+use EmailLog\Core\UI\Setting\SettingField;
+
+class EmailLogSetting extends Setting {
+
+	/**
```
Comment (sudar) [inline include/Core/UI/Setting/EmailLogSetting.php]: The phpdoc block for this method and the other ones in this class should be changed.

File: include/Core/UI/Setting/EmailLogSetting.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	 * Implement `initialize()` method.
+	 */
+	protected function initialize() {
+		$this->section->id          = 'email-log';
+		$this->section->title       = __( 'Email Log Settings', 'email-log' );
+		$this->section->option_name = 'el_email_log';
```
Comment (sudar) [inline include/Core/UI/Setting/EmailLogSetting.php]: Let's change this to `el_email_log_core`

File: include/Core/UI/Setting/EmailLogSetting.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	 * Implement `get_fields()` method.
+	 *
+	 * @return array
+	 */
+	public function get_fields() {
+		$fields = array();
+
+		$email_log_fields = array(
+			'allowed_user_roles' => __( 'Allowed User Roles', 'email-log' ),
+			'remove_email_logs'  => __( 'Remove Data on Uninstall?', 'email-log' ),
```
Comment (sudar) [inline include/Core/UI/Setting/EmailLogSetting.php]: Let's rename this to `remove_on_uninstall`

File: include/Core/UI/Setting/EmailLogSetting.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		}
+		return $values;
+	}
+
+	/**
+	 * Renders the Email Log settings fields.
+	 *
+	 * @param array $args
+	 */
+	public function render_email_log_settings( $args ) {
```
Comment (sudar) [inline include/Core/UI/Setting/EmailLogSetting.php]: Split this up into two methods and assign them to each `SettingField`

File: include/Core/UI/Setting/EmailLogSetting.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+	/**
+	 * Checks the Checkbox when values are present in a given array.
+	 *
+	 * Use this function in Checkbox fields.
+	 *
+	 * @param array $values   List of all possible values.
+	 * @param string $current The current value to be checked.
+	 */
+	public function checked_array( $values, $current ) {
```
Comment (sudar) [inline include/Core/UI/Setting/EmailLogSetting.php]: Let's make this as a util function and move it to Util namespace

File: include/Util/helper.php [hunk 6]
```diff
@@ -72,4 +72,32 @@ function get_log_columns_to_export() {
...
+}
+
+/**
+ * Returns TRUE if the User is Administrator or the User's role is allowed in Plugin's settings page.
+ *
+ * @since 2.1.0
+ *
+ * @return bool
+ */
+function can_current_user_email_log() {
```
Comment (sudar) [inline include/Util/helper.php]: Let's rename this to `can_current_user_view_email_log`

File: include/Util/helper.php [hunk 7]
```diff
@@ -72,4 +72,32 @@ function get_log_columns_to_export() {
...
+	$return_value = false;
+	$option       = get_option( 'el_email_log' );
+
+	if ( current_user_can( 'administrator' ) ) {
```
Comment (sudar) [inline include/Util/helper.php]: Let's use guard conditions to make this method more readable.

You can read more about guard conditions at https://carlalexander.ca/mastering-php-conditionals/

File: include/Core/EmailLog.php [hunk 8]
```diff
@@ -65,6 +65,14 @@ class EmailLog {
 	public $table_manager;
 
 	/**
+	 * Email Log Plugins Setting.
+	 *
+	 * @since 2.1.0
+	 * @var \EmailLog\Core\UI\Setting\Setting
+	 */
+	public $setting;
```
Comment (sudar) [inline include/Core/EmailLog.php]: Do we need to reference this variable anywhere? If no, then set this as an `loadie`

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 2 above)
Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: Fixed.

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 3 above)
Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: Fixed

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 4 above)
Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: Fixed.

Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: Do you want me to isolate `sanitize()` function for each Setting field?

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 5 above)
Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: Moved.

File: include/Util/helper.php (hunk 6 above)
Comment (mariadanieldeepak) [reply include/Util/helper.php]: Fixed.

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 4 above)
Comment (sudar) [reply include/Core/UI/Setting/EmailLogSetting.php]: For now, let's not change `sanitize`. But it needs to be changed eventually.

File: include/Core/EmailLog.php (hunk 8 above)
Comment (mariadanieldeepak) [reply include/Core/EmailLog.php]: I'm sorry, I didn't get this. Should this variable be moved to interface `loadie`?

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 1 above)
Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: Do you mean to add the Doc block only to the Class here?

File: include/Core/EmailLog.php (hunk 8 above)
Comment (sudar) [reply include/Core/EmailLog.php]: What I meant is to remove the variable and then pass a reference to the settings object using the `add_loadie` function.

That way we can remove a member variable in the class and make it much more decoupled.

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 1 above)
Comment (sudar) [reply include/Core/UI/Setting/EmailLogSetting.php]: Yeah. If the methods in the parent class already have a phpdoc block and in the child class we are just inheriting the class without any additional functionality (especially in the case of abstract methods) then we can ignore the phpdoc block for the inherited methods.

phpdoc will automatically inherit documentation.

We need to add phpdoc block in inherited methods only if we are going to have a different phpdoc block or if we have to add any child class specific information.

File: include/Core/EmailLog.php (hunk 8 above)
Comment (mariadanieldeepak) [reply include/Core/EmailLog.php]: > What I meant is to remove the variable and then pass a reference to the settings object using the `add_loadie` function.

Perfect! Just understood.

Comment (mariadanieldeepak) [reply include/Core/EmailLog.php]: > That way we can remove a member variable in the class and make it much more decoupled.

Perfect, just one line change, but gives an entirely different meaning. Fixed it now.

File: include/Core/UI/Setting/EmailLogSetting.php (hunk 1 above)
Comment (mariadanieldeepak) [reply include/Core/UI/Setting/EmailLogSetting.php]: > We need to add phpdoc block in inherited methods only if we are going to have a different phpdoc block or if we have to add any child class specific information.

Understood!

**Pull request merged**